compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -368,6 +419,135 @@ void LinalgFusePass::runOnOperation() {
     return signalPassFailure();
   }
 }
+namespace {
+/// Pattern to wrap splitReduction transformation.
+struct CodegenSplitReduction
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+  CodegenSplitReduction(MLIRContext *context, bool reductionReordering,
+                        int64_t size, uint64_t unrollFactor,
+                        linalg::LinalgTransformationFilter filter,
+                        PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
+        reductionReordering(reductionReordering),
+        size(size),
+        unrollFactor(unrollFactor),
+        filter(std::move(filter)) {}
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that only generic ops are supported, only one input and the
+    // innermost tensor dimension is reduction dimension.  Everything else is
+    // copied from the original SplitReduction Transformation.
+    // TODO: support named ops, numInputs > 1, and modify lastDim check below
+    // accordingly. If reductionReordering is not enabled by default, it must be
+    // an integer or index type to proceed to allow associative reordering.
+    if (failed(filter.checkAndNotify(rewriter, op)) ||
+        !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
+        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations() ||
+        !isa<linalg::GenericOp>(op) || op.getNumInputs() != 1 ||
+        !(reductionReordering || op.getInputOperand(0)
+                                     ->get()
+                                     .getType()
+                                     .dyn_cast<ShapedType>()
+                                     .getElementType()
+                                     .isIntOrIndex()))
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+    SmallVector<unsigned> dims;
+    op.getReductionDims(dims);
+    AffineMap map = op.getMatchingIndexingMap(op.getInputOperand(0));
+    unsigned lastIdx = map.getNumResults() - 1;
+    unsigned lastDim = map.getDimPosition(lastIdx);
+    if (lastDim != dims[0]) {
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+    }
+
```

Comment:
> My understanding is that we'd like to check if the last dimension is reduction. We should derive the information from the Linalg op generally because it's how structure ops (i.e., linalg ops) work. We don't need to build the condition based on the precondition of hasOneInput, hasExactlyOneReduction, etc.

No, we have to check that the last rank is reduction dimension.